tools/pd-api-bench/main.go

```diff
@@ -0,0 +1,225 @@
+	// qps
+	qps = flag.Int("qps", 1000, "qps")
+	// tls
+	caPath   = flag.String("cacert", "", "path of file that contains list of trusted SSL CAs")
+	certPath = flag.String("cert", "", "path of file that contains X509 certificate in PEM format")
+	keyPath  = flag.String("key", "", "path of file that contains X509 key in PEM format")
+)
+
+var base = int(time.Second) / int(time.Millisecond)
+
+func main() {
+	flag.Parse()
+	ctx, cancel := context.WithCancel(context.Background())
+	sc := make(chan os.Signal, 1)
+	signal.Notify(sc,
+		syscall.SIGHUP,
+		syscall.SIGINT,
+		syscall.SIGTERM,
+		syscall.SIGQUIT)
+	var sig os.Signal
+	go func() {
+		sig = <-sc
+		cancel()
+	}()
+
+	if *minResolvedTSByGRPC {
+		go handleMinResolvedTSByGRPC(ctx)
+	}
+	if *minResolvedTSByHTTP {
+		go handleMinResolvedTSByHTTP(ctx)
+	}
+
+	<-ctx.Done()
+	switch sig {
+	case syscall.SIGTERM:
+		exit(0)
+	default:
+		exit(1)
+	}
+}
+
+func handleMinResolvedTSByGRPC(ctx context.Context) {
+	log.Println("handleMinResolvedTSByGRPC start...")
+	tt := base / *qps
+	for i := 0; i < *client; i++ {
+		pdCli := newPDClient()
+		go func() {
+			var ticker = time.NewTicker(time.Millisecond * time.Duration(tt))
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
+					_, _, err := pdCli.GetMinResolvedTS(ctx, []uint64{1, 2, 3})
+					if err != nil {
+						log.Println(err)
+						continue
+					}
+				case <-ctx.Done():
+					log.Println("Got signal to exit handleMinResolvedTSByGRPC")
+					return
+				}
+			}
+		}()
+	}
+}
+
+func handleMinResolvedTSByHTTP(ctx context.Context) {
+	log.Println("handleMinResolvedTSByHTTP start...")
+	// Judge whether with tls.
+	protocol := "http"
+	if len(*caPath) != 0 {
+		protocol = "https"
+	}
+	url := fmt.Sprintf("%s://%s/pd/api/v1/min-resolved-ts", protocol, *pdAddr)
+
+	tt := base / *qps
+	for i := 0; i < *client; i++ {
+		httpsCli := newHttpClient()
+		go func() {
+			// Mock client-go's request frequency.
+			var ticker = time.NewTicker(time.Millisecond * time.Duration(tt))
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
+					storeID := 1
+					req, _ := http.NewRequest(http.MethodGet, fmt.Sprintf("%s/%d", url, storeID), nil)
+					res, err := httpsCli.Do(req)
+					if err != nil {
+						log.Println("error: ", err)
+						continue
+					}
+					listResp := &minResolvedTS{}
+					apiutil.ReadJSON(res.Body, listResp)
+					res.Body.Close()
+				case <-ctx.Done():
+					log.Println("Got signal to exit handleMinResolvedTSByHTTP")
+					return
+				}
+			}
+		}()
+	}
+}
+
+type minResolvedTS struct {
+	IsRealTime          bool              `json:"is_real_time,omitempty"`
+	MinResolvedTS       uint64            `json:"min_resolved_ts"`
+	PersistInterval     typeutil.Duration `json:"persist_interval,omitempty"`
+	StoresMinResolvedTS map[uint64]uint64 `json:"stores_min_resolved_ts"`
+}
+
+// newHttpClient returns an HTTP(s) client.
+func newHttpClient() *http.Client {
```

Comment:
why the pd server cannot reuse the client?

Reply:
This is because I want to simulate the http approach in which `a tidb` corresponds to `a new client`.

Reply:
I think the problem is no need to use gRPC to resolve, gRPC is also based on http2. we should to find why HTTP connect cannot reuse.

Reply:
The test results show that it's because of this https://github.com/tikv/pd/pull/6914